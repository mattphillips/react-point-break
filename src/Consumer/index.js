import React from 'react';
import PropTypes from 'prop-types';

import Breakpoint from '../Breakpoint';
import { Consumer, defaultValue } from '../Context';

const BreakpointConsumer = ({ children, render, ...queries }) => {
  const renderFunction = render !== undefined ? render : children;
  return (
    <Consumer>
      {value => {
        const providerQueries = value === defaultValue ? {} : value;
        return <Breakpoint queries={{ ...providerQueries, ...queries }}>{renderFunction}</Breakpoint>;
      }}
    </Consumer>
  );
};

BreakpointConsumer.propTypes = {
  children: PropTypes.func,
  render: PropTypes.func
};

export default BreakpointConsumer;
