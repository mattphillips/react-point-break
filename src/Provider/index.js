import React from 'react';
import PropTypes from 'prop-types';

import { Provider } from '../Context';

const BreakPointProvider = ({ children, ...queries }) => {
  return <Provider value={queries}>{children}</Provider>;
};

BreakPointProvider.propTypes = {
  children: PropTypes.element.isRequired
};

export default BreakPointProvider;
