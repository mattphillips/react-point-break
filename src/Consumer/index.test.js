import React from 'react';
import { mount } from 'enzyme';

import Provider from '../Provider';
import Consumer from './';

const noop = () => {};
const queries = {
  md: '(min-width: 801px) and (max-width: 900px)',
  lg: '(min-width: 901px)'
};

describe('<Consumer />', () => {
  it('renders given children render function', () => {
    const component = mount(<Consumer>{() => <h1>Hello, world!</h1>}</Consumer>);
    expect(component).toMatchSnapshot();
  });

  it('renders Consumer with given queries', () => {
    window.matchMedia = () => ({ addListener: noop });

    const component = mount(
      <Consumer {...queries}>
        {({ md, lg }) => {
          if (lg) return <h1>Large</h1>;
          if (md) return <h1>Medium</h1>;
          return <h1>Small</h1>;
        }}
      </Consumer>
    );
    expect(component).toMatchSnapshot();
  });

  it('renders Consumer with provider queries and given queries', () => {
    window.matchMedia = () => ({ addListener: noop });

    const component = mount(
      <Provider sm="small query">
        <Consumer {...queries}>
          {({ sm, md, lg }) => {
            if (lg) return <h1>Large</h1>;
            if (md) return <h1>Medium</h1>;
            if (sm) return <h1>Small</h1>;
            return <h1>Extra Small</h1>;
          }}
        </Consumer>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it('overrides matching provider queries with given consumer queries', () => {
    window.matchMedia = () => ({ addListener: noop });

    const component = mount(
      <Provider {...queries}>
        <Consumer lg="large override" md="medium override">
          {({ md, lg }) => {
            if (lg) return <h1>Large</h1>;
            if (md) return <h1>Medium</h1>;
            return <h1>Small</h1>;
          }}
        </Consumer>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
