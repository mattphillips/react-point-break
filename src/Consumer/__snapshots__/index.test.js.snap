// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Consumer /> overrides matching provider queries with given consumer queries 1`] = `
<BreakPointProvider
  lg="(min-width: 901px)"
  md="(min-width: 801px) and (max-width: 900px)"
>
  <Provider
    value={
      Object {
        "lg": "(min-width: 901px)",
        "md": "(min-width: 801px) and (max-width: 900px)",
      }
    }
  >
    <BreakpointConsumer
      lg="large override"
      md="medium override"
    >
      <Consumer>
        <Breakpoint
          queries={
            Object {
              "lg": "large override",
              "md": "medium override",
            }
          }
        >
          <h1>
            Small
          </h1>
        </Breakpoint>
      </Consumer>
    </BreakpointConsumer>
  </Provider>
</BreakPointProvider>
`;

exports[`<Consumer /> renders Consumer with given queries 1`] = `
<BreakpointConsumer
  lg="(min-width: 901px)"
  md="(min-width: 801px) and (max-width: 900px)"
>
  <Consumer>
    <Breakpoint
      queries={
        Object {
          "lg": "(min-width: 901px)",
          "md": "(min-width: 801px) and (max-width: 900px)",
        }
      }
    >
      <h1>
        Small
      </h1>
    </Breakpoint>
  </Consumer>
</BreakpointConsumer>
`;

exports[`<Consumer /> renders Consumer with provider queries and given queries 1`] = `
<BreakPointProvider
  sm="small query"
>
  <Provider
    value={
      Object {
        "sm": "small query",
      }
    }
  >
    <BreakpointConsumer
      lg="(min-width: 901px)"
      md="(min-width: 801px) and (max-width: 900px)"
    >
      <Consumer>
        <Breakpoint
          queries={
            Object {
              "lg": "(min-width: 901px)",
              "md": "(min-width: 801px) and (max-width: 900px)",
              "sm": "small query",
            }
          }
        >
          <h1>
            Extra Small
          </h1>
        </Breakpoint>
      </Consumer>
    </BreakpointConsumer>
  </Provider>
</BreakPointProvider>
`;

exports[`<Consumer /> renders given children render function 1`] = `
<BreakpointConsumer>
  <Consumer>
    <Breakpoint
      queries={Object {}}
    >
      <h1>
        Hello, world!
      </h1>
    </Breakpoint>
  </Consumer>
</BreakpointConsumer>
`;

exports[`<Consumer /> renders given render function 1`] = `
<BreakpointConsumer
  render={[Function]}
>
  <Consumer>
    <Breakpoint
      queries={Object {}}
    >
      <h1>
        Hello, world!
      </h1>
    </Breakpoint>
  </Consumer>
</BreakpointConsumer>
`;
